AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -703,6 +704,69 @@ - (instancetype)styledWithBlock:(void (^)(ASLayoutElementStyle *style))styleBloc
 
 #pragma mark - Layout
 
+- (void)invalidateSize
+{
+  ASDisplayNodeAssertThreadAffinity(self);
+  
+  __instanceLock__.lock();
+
+  // Mark the node for layout in the next layout pass
+  [self setNeedsLayout];
+  
+  ASDisplayNode *supernode = _supernode;
+  if (supernode) {
+    __instanceLock__.unlock();
+    // Cause supernode's layout to be invalidated
+    // We need to release the lock to prevent a deadlock
+    [supernode invalidateSize];
+    return;
+  }
+  
+  // We are the root node and need to re-flow the layout; one of our children requested to have its size re-set.
+  CGSize oldSize = self.bounds.size;
+
+  // Figure out constrainedSize to use
+  ASSizeRange constrainedSize = ASSizeRangeMake(self.bounds.size);
+  if (_pendingDisplayNodeLayout != nullptr) {
+    constrainedSize = _pendingDisplayNodeLayout->constrainedSize;
+  } else if (_calculatedDisplayNodeLayout->layout != nil) {
+    constrainedSize = _calculatedDisplayNodeLayout->constrainedSize;
+  }
+
+  // Check if the returned layout has a different size as the current bounds
+  ASLayout *layout = [self layoutThatFits:constrainedSize];
+  if (CGSizeEqualToSize(oldSize, layout.size) == NO) {
+    // If the size of the layout changes inform our container (e.g ASTableView, ASCollectionView, ASViewController, ...)
+    // that we need it to change our bounds size.
+    [self displayNodeDidInvalidateSizeNewSize:layout.size];
```

Comment:
I see.  I think we could REQUIRE_SUPER on that method and then the base class would still be called?

Since layoutThatFits is indeed applying it to the calculatedLayout, wouldn't it be correct from an API sense to call -calculatedLayoutDidChange even if we use a different internal method too?

We should note a followup for the names of _pending and _calculated layout - it could require a tweak in API naming.  Something like _pending and _appliedLayout might be better, or even _calculated and _applied.  Then there could be a difference between appliedLayoutDidChange (maybe this is just -layoutDidFinish?) and calculatedLayoutDidChange.  If we had that distinction, then yes definitely a separate internal method seems right for this use case.